L54-python-server-wait.md

```diff
@@ -0,0 +1,126 @@
+gRPC Python Server Wait API
+----
+* Author(s): lidiz
+* Approver: gnossen
+* Status: Draft
+* Implemented in: Python
+* Last updated: June 12, 2019
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/GhJl_JHfaHA
+
+## Abstract
+
+Add a `wait_for_termination` API for Python server that can block in any thread
+until the server is stopped or terminated.
+
+## Background
+
+Currently, the gRPC Python server only provides a `server.Start()` API, but
+there isn't any API to block forever. gRPC Python server applications need to
+block on the main thread to keep serving explicitly. The solution the gRPC
+Python team proposed in examples is sleeping for a while then exit. See code:
+
+```Python
+_ONE_DAY_IN_SECONDS = 60 * 60 * 24
+
+def serve():
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+    route_guide_pb2_grpc.add_RouteGuideServicer_to_server(
+        RouteGuideServicer(), server)
+    server.add_insecure_port('[::]:50051')
+    server.start()
+    try:
+        while True:
+            time.sleep(_ONE_DAY_IN_SECONDS)
+    except KeyboardInterrupt:
+        server.stop(0)
+```
+
+Developers are running server longer than one day, and hope it to keep running
+forever if possible. gRPC users are coming up with workarounds to block the main
+thread:
+
+1) Use other API to block like `select.select`;
+2) Sleep longer.
+
+Ideally, this feature should be part of gRPC library, because only gRPC library
+have the clearest signal when the server is going to shutdown. This feature
+existed in C++ as
+[`Wait()`](https://grpc.github.io/grpc/cpp/classgrpc_1_1_server_interface.html#ac36477b6a7593a4e4608c7eb712b0d70),
+in Java as `blockUntilShutdown()`, in Golang as `Serve`. This gRFC is
+introducing the Python version of it.
+
+### Related Proposals: 
+
+None.
+
+## Proposal
+
+This gRFC propose to add following API to `grpc.Server` interface:
+
+```Python
+class Server:
+
+    def wait_for_termination(self):
+        """Block current thread until the server stops.
+        
+        This is an EXPERIMENTAL API.
+        
+        The wait will not consume computational resources during blocking, and it
+        will block indefinitely. There are two ways to unblock:
+        
+        1) Calling `stop` on the server in another thread;
+        2) The `__del__` of the server object is invoked.
+        """
+        raise NotImplementedError()
+```
+
+## Rationale
+
+### Alternative Design: Await For Signal
```

Comment:
I definitely think the way you chose is the way to go. The application author should be in charge of what runs on which thread. They can write their own signal handler to run on the main thread and call `server.stop()`.

Reply:
Glad to see that we have similar view of how this API should be done :)
I added another alternative design section describing the idea proposed by @mehrdada in the implementation PR.